### ISSUE TITLE
Out of memory in case of Splunk indexer slowness/failure

### ISSUE THREAD

**Issue by ludovic-boutros** [splunk/kafka-connect-splunk#423]
Hello, 
We are using the Splunk Sink Connector with these main parameters:

```json
{
	"name": "SplunkHECSinkConnector",
	"config":{
		"connector.class": "com.splunk.kafka.connect.SplunkSinkConnector",
		"tasks.max": "6",
		"splunk.hec.ack.enabled": "true",
		"splunk.hec.max.outstanding.events": "50000",
		"splunk.hec.max.retries": "-1",
		"splunk.hec.backoff.threshhold.seconds": "60",
		"splunk.hec.threads": "1"
	}
}
```

In my understanding, we should never have more than 50000 events per task kept in memory.
But that is not the case if Splunk indexers encounter slowness or failures.

We can observe in the Kafka Connect logs such errors and messages:

```shell
[2024-02-27 06:39:24,527] INFO [SplunkHECSinkConnector|task-5] handled 394 failed batches with 193452 events (com.splunk.kafka.connect.SplunkSinkTask:154)
```

I have attached the Kafka Connect metrics during a Splunk indexer stress test.
You can observe the out of memory and the number of active records.

<img width="861" alt="out-of-memory-splunk-sink" src="https://github.com/splunk/kafka-connect-splunk/assets/13764888/7c210628-71c0-4b12-be9a-1fd25f5205ff">

**Comment by VihasMakwana** [comment]
@ludovic-boutros thanks for this.
I will take a look and give my thoughts on this.

**Comment by VihasMakwana** [comment]
@ludovic-boutros can you attach the entire kafka connect logs?

**Comment by ludovic-boutros** [comment]
@VihasMakwana we will open a case on Splunk side in order to send you the complete logs in a more secure way.
We decreased the outstanding event property to 10000 and still we have OOM issues.

![image](https://github.com/splunk/kafka-connect-splunk/assets/13764888/c3083f94-b55f-48f6-aacb-2e17f0be9328)
![image](https://github.com/splunk/kafka-connect-splunk/assets/13764888/0da1d4e5-e46d-40f7-8bc3-f7aabbee8ef3)

**Comment by ludovic-boutros** [comment]
I still don't understand how the `sink-record-active-count` can be so high? Shouldn't it stay under the `splunk.hec.max.outstanding.events` value?

**Comment by ludovic-boutros** [comment]
I have put some classes in debug log level.
Here is on interesting one:
```
[2024-04-09 10:54:52,049] DEBUG [SplunkHECSinkIntConnector|task-0] tid=152 received 261 records with total -4656978 outstanding events tracked (com.splunk.kafka.connect.SplunkSinkTask:83)
```
I don't think this negative number is normal ;)

**Comment by ludovic-boutros** [comment]
This incorrect outstanding event count makes the outstanding event limit ineffective and this leads to an out of memory in case of Splunk slowness/failure. This is my understanding.

**Comment by ludovic-boutros** [comment]
@VihasMakwana I did not manage to understand how the outstanding event count could be lower than zero.
Do you think we could just add a small check and keep it greater than or equal to zero in the event tracker?
I don't know if it would add side effects, but it would at least assert that the max outstanding event count would be effective.

**Comment by ludovic-boutros** [comment]
@VihasMakwana I have patched the connector to prevent negative event count and we can see the effect on the number of events kept in memory. It does not fix the real issue but at least the "symptoms".
![image](https://github.com/splunk/kafka-connect-splunk/assets/13764888/19d0baf8-acd8-4877-93dc-38ccc0870b90)

**Comment by ludovic-boutros** [comment]
Resolved by #431
